crucible-llvm/src/Lang/Crucible/LLVM/SimpleLoopFixpointCHC.hs

```diff
@@ -0,0 +1,1415 @@
+  CallFrameContext sym wptr ext init ret blocks
+unwrapSomeCallFrameContext blocks_repr (SomeCallFrameContext ctx) =
+  case W4.testEquality blocks_repr (fmapFC C.blockInputs $ C.cfgBlockMap $ callFrameContextCFG ctx) of
+    Just Refl -> ctx
+    Nothing -> C.panic "SimpleLoopFixpoint.unwrapSomeCallFrameContext" ["type mismatch"]
+
+data ExecutionFeatureContext sym wptr ext = ExecutionFeatureContext
+  { executionFeatureContextFixpointStates :: C.FnHandleMap (SomeCallFrameContext sym wptr ext)
+  , executionFeatureContextInvPreds :: [W4.SomeSymFn sym]
+  , executionFeatureContextLoopFunEquivConds :: [W4.Pred sym]
+  }
+
+callFrameContextLookup ::
+  CallFrameHandle init ret blocks ->
+  ExecutionFeatureContext sym wptr ext ->
+  CallFrameContext sym wptr ext init ret blocks
+callFrameContextLookup (CallFrameHandle hdl blocks_repr) ctx =
+  unwrapSomeCallFrameContext blocks_repr $
+    fromMaybe (C.panic "SimpleLoopFixpoint.callFrameContextLookup" ["missing call frame context", show hdl]) $
+      C.lookupHandleMap hdl (executionFeatureContextFixpointStates ctx)
+
+callFrameContextUpdate ::
+  CallFrameHandle init ret blocks ->
+  (CallFrameContext sym wptr ext init ret blocks -> CallFrameContext sym wptr ext init ret blocks) ->
+  ExecutionFeatureContext sym wptr ext ->
+  ExecutionFeatureContext sym wptr ext
+callFrameContextUpdate (CallFrameHandle hdl blocks_repr) f ctx =
+  ctx
+    { executionFeatureContextFixpointStates = C.updateHandleMap
+        (SomeCallFrameContext . f . unwrapSomeCallFrameContext blocks_repr)
+        hdl
+        (executionFeatureContextFixpointStates ctx)
+    }
+
+callFrameContextLookup' ::
+  CallFrameHandle init ret blocks ->
+  C.BlockID blocks args ->
+  ExecutionFeatureContext sym wptr ext ->
+  Maybe (FixpointState sym wptr blocks args)
+callFrameContextLookup' hdl bid ctx =
+  MapF.lookup bid $ callFrameContextFixpointStates $ callFrameContextLookup hdl ctx
+
+callFrameContextInsert ::
+  CallFrameHandle init ret blocks ->
+  C.BlockID blocks args ->
+  FixpointState sym wptr blocks args ->
+  ExecutionFeatureContext sym wptr ext ->
+  ExecutionFeatureContext sym wptr ext
+callFrameContextInsert hdl bid fs =
+  callFrameContextUpdate hdl $
+    \ctx -> ctx { callFrameContextFixpointStates = MapF.insert bid fs (callFrameContextFixpointStates ctx) }
+
+callFrameContextPush ::
+  CallFrameHandle init ret blocks ->
+  C.Some (C.BlockID blocks) ->
+  ExecutionFeatureContext sym wptr ext ->
+  ExecutionFeatureContext sym wptr ext
+callFrameContextPush hdl bid =
+  callFrameContextUpdate hdl $
+    \ctx -> ctx { callFrameContextLoopHeaders = bid : callFrameContextLoopHeaders ctx }
+
+callFrameContextPop ::
+  CallFrameHandle init ret blocks ->
+  ExecutionFeatureContext sym wptr ext ->
+  ExecutionFeatureContext sym wptr ext
+callFrameContextPop  hdl =
+  callFrameContextUpdate hdl $
+    \ctx -> ctx { callFrameContextLoopHeaders = tail $ callFrameContextLoopHeaders ctx }
```

Comment:
Since `tail` is partial, do we have any protection that we won't enter here without frame context loop headers?  Maybe a `panic` in that case would be helpful (although GHC 9.10's exception backtraces may finally make these easier to track down).